src/lib/runtime/mounts/dev/dev.c

```diff
@@ -108,59 +108,81 @@ int _singularity_runtime_mount_dev(void) {
             singularity_message(ERROR, "Failed to mount %s: %s\n", joinpath(devdir, "/shm"), strerror(errno));
             ABORT(255);
         }
+        singularity_priv_drop();
 
         if ( singularity_config_get_bool_char(MOUNT_DEVPTS) > 0 ) {
             struct group *ttygid;
-            char *devpts_opts_base="mode=0620,newinstance,ptmxmode=0666,gid=";
+            char *devpts_opts_base="mode=0620,newinstance,ptmxmode=0666";
+            char *devpts_opts_gid=",gid=";
             char *devpts_opts;
-            unsigned int max_sz, gd, gd_n;
+            unsigned int max_sz=0, gd, gd_n;
 
-            if ( (ttygid=getgrnam("tty")) == NULL ) {
-                singularity_message(ERROR, "Problem resolving 'tty' group GID: %s\n", strerror(errno));
-                ABORT(255);
-            }
+            if ( singularity_suid_enabled() > 0 ) {
 
-            /* number of digits in gid */
-            if ( ttygid->gr_gid == 0 ) {
-                gd_n = 1;
-            }
-            else {
-                gd_n = 0;
-                gd = ttygid->gr_gid;
-                while ( gd ) {
-                    gd /= 10;
-                    gd_n++;
+                if ( (ttygid=getgrnam("tty")) == NULL ) {
+                    singularity_message(ERROR, "Problem resolving 'tty' group GID: %s\n", strerror(errno));
+                    ABORT(255);
                 }
-            }
 
-            /* length of gid string + mount options + terminator + padding */
-            max_sz = gd_n + strlen(devpts_opts_base) + 1 + 16;
-            if ( (devpts_opts=malloc(max_sz)) == NULL ) {
-                    singularity_message(ERROR, "Memory allocation failure: %s\n", strerror(errno));
-                    ABORT(255);
+                /* number of digits in gid */
+                if ( ttygid->gr_gid == 0 ) {
+                    gd_n = 1;
+                }
+                else {
+                    gd_n = 0;
+                    gd = ttygid->gr_gid;
+                    while ( gd ) {
+                        gd /= 10;
+                        gd_n++;
+                    }
+                }
+
+                /* length of gid string + mount options + terminator + padding */
+                max_sz = gd_n + strlen(devpts_opts_base) + strlen(devpts_opts_gid) + 1 + 16;
+                if ( (devpts_opts=malloc(max_sz)) == NULL ) {
+                        singularity_message(ERROR, "Memory allocation failure: %s\n", strerror(errno));
+                        ABORT(255);
+                }
+                bzero(devpts_opts, max_sz);
+                snprintf(devpts_opts, max_sz-1, "%s%s%d", devpts_opts_base, devpts_opts_gid, ttygid->gr_gid);
+            }
+            else {
+                singularity_message(DEBUG, "Setuid not allowed - not setting /dev/pts filesystem gid\n");
```

Comment:
@hollowec This message `Setuid not allowed` is confusing, when I read it I thought I was in the setuid codepath, maybe change into `Not setting /dev/pts filesystem gid: user namespace enabled` to avoid confusion